xtask/src/semver_check.rs

```diff
@@ -0,0 +1,257 @@
+    // version of `rustdoc_types` here
+    //
+    // Hopefully this whole pre-processing is just a stop-gap solution until it's
+    // possible to generate docs for the stable-API only.
+
+    log::info!("{:?}", path);
+    let json_string = std::fs::read_to_string(path)?;
+    let mut krate: rustdoc_types::Crate = serde_json::from_str(&json_string)?;
+
+    let mut to_remove = vec![];
+
+    // first pass - just look for cfg-gated items
+    //
+    // the string to match depends on the rustfmt-json version!
+    // later version emit `#[<cfg>(...` instead
+    let cfg_gates = vec![
+        "#[cfg(any(doc, feature = \"unstable\"))]",
+        "#[cfg(feature = \"unstable\")]",
+    ];
+
+    for (id, item) in &mut krate.index {
+        if item
+            .attrs
+            .iter()
+            .any(|attr| cfg_gates.contains(&attr.as_str()))
+        {
+            to_remove.push(id.clone());
+        }
+    }
+
+    log::debug!("Items to remove {:?}", to_remove);
+
+    for id in &to_remove {
+        krate.index.remove(&id);
+        krate.paths.remove(&id);
+    }
+
+    for (_id, item) in &mut krate.index {
+        match &mut item.inner {
+            ItemEnum::Module(module) => {
+                module.items = module
+                    .items
+                    .iter()
+                    .filter(|id| !to_remove.contains(id))
+                    .map(|id| id.clone())
+                    .collect();
+            }
+            ItemEnum::Struct(struct_) => {
+                struct_.impls = struct_
+                    .impls
+                    .iter()
+                    .filter(|id| !to_remove.contains(id))
+                    .map(|id| id.clone())
+                    .collect();
+
+                struct_.impls = struct_
+                    .impls
+                    .iter()
+                    .filter(|id| !to_remove.contains(id))
+                    .map(|id| id.clone())
+                    .collect();
+
+                match &mut struct_.kind {
+                    rustdoc_types::StructKind::Unit => (),
+                    rustdoc_types::StructKind::Tuple(_ids) => (),
+                    rustdoc_types::StructKind::Plain {
+                        fields,
+                        has_stripped_fields: _,
+                    } => {
+                        for id in &to_remove {
+                            if let Some(found) = fields.iter().enumerate().find(|i| i.1 == id) {
+                                fields.remove(found.0);
+                            }
+                        }
+                    }
+                }
+
+                if struct_.impls.is_empty() {
+                    to_remove.push(_id.clone());
+                }
+            }
+            ItemEnum::Enum(enum_) => {
+                enum_.impls = enum_
+                    .impls
+                    .iter()
+                    .filter(|id| !to_remove.contains(id))
```

Comment:
Are you looking for [`Vec::retain`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.retain)?